src/librustc_span/span_encoding.rs

```diff
@@ -65,13 +65,24 @@ use std::marker::{StructuralEq, StructuralPartialEq};
 #[derive(Clone, Copy)]
 #[repr(C)] // So we can transmute to a `u64`.
 pub struct Span {
-    base_or_index: u32,
+    // We can transmute `SpanBaseOrIndex` to an `u32` as it is internally
+    // represented as a scalar value.
+    base_or_index: SpanBaseOrIndex,
     len_or_tag: u16,
     ctxt_or_zero: u16,
 }
 
+rustc_index::newtype_index! {
+    pub struct SpanBaseOrIndex {
+        const DUMMY_SPAN_INDEX = 0,
+    }
+}
+
+#[cfg(target_arch = "x86_64")]
```

Comment:
This is worth a comment. E.g. "`SpanBaseOrIndex` is implemented via `newtype_index`, which means that `Option<Span>` takes the same space as `Span`.